### DISCUSSION TITLE
How to set a custom background image?

### DISCUSSION THREAD

**Discussion by davekimi14** [ipxe/ipxe#945] [Q&A] [answered]
Hi Everyone!
I had a quick question. How would I specify a path to a custom .png image that is located on my local device to be used as a background image for the menu. I already enabled the 2 libraries required for background images and I can set images form the web, but how do I do this from my local machine. And the second one is what are the console/image size limitations/restrictions. 
Thank you!

**Comment by NiKiZe** [comment]
Limitations depends on your video device or firmware/vgabios.

In efi mode you can use file:// to access local resources, but it's support is limited.

Are you booting iPXE as well from local device?

**Reply by davekimi14** [reply]
I am booting multiple machines from the same network, so no not the local device. 
I'm guessing since I'm doing that the file:// prefix won't apply here? Do I have any other options other than uploading this image to a http server?

**Reply by stappersg** [reply]
Please elaborate how to match
> Do I have any other options other than uploading this image to a http server?

with _"a custom .png image that is located on my local device"_ from the original question.

**Reply by NiKiZe** [reply]
So what do you mean with "local device" from
> How would I specify a path to a custom .png image that is located on my local device

?

**Reply by davekimi14** [reply]
Sorry, by local device I mean the server running ipxe

**Reply by NiKiZe** [reply]
iPXE tunna on the local client machine.
The "server" usually consists of DHCP+TFTP services and maybe a webserver.

You would load the image as any other file, see https://ipxe.org/cmd/imgfetch for a good general description.

So if you have any chain, kernel or initrd line in your script, loading the image would work the same.

When trying to load it, check your server logs for any errors regarding locating the file.

Also when asking for help it is often a good idea to include a minimal script, showing what you are trying to do, and what you have so far.

**Reply by davekimi14** [reply]
Hi!
I'm including a screenshot of my boot.ipxe file. I am trying to set a .png file as my background. This background image is on my tftp/ipxe host. I am wondering if I could use this image without having to upload it to an http server. 
![GIS-AVD_2 1 2023-04-23 16-34-53](https://user-images.githubusercontent.com/46170802/233872578-c29a450d-302e-4df9-aa93-612aba698243.png)

**Reply by davekimi14** [reply]
> Hi!
> I'm including a screenshot of my boot.ipxe file. I am trying to set a .png file as my background. This background image is on my tftp/ipxe host. I am wondering if I could use this image without having to upload it to an http server.
> ![GIS-AVD_2 1 2023-04-23 16-34-53](https://user-images.githubusercontent.com/46170802/233872578-c29a450d-302e-4df9-aa93-612aba698243.png)

So I changed the file:// to tftp:// and now I in the boot up sequence I see that the tftp file loads "ok" and the next line after that says "Could not configure console: No such file or directory"

**Reply by stappersg** [reply]
Things to do better:
- Try to share text also as text
- Try to share error messages also as a picture

Things to try:
- Change `tftp://var/tftp/menu/logo.png` to `tftp://hostname/menu/logo.png` or just `menu/logo.png`.
- Understand what the client, _iPXE_, asks from the (TFTP) server.
- Understand how the client should address the server.

**Reply by NiKiZe** [reply]
You can load it over TFTP (or any other supported protocol) as well, http is not a requirement.

If you have text... Don upload it as an image. Why are you not copy pasting the text?

**Reply by davekimi14** [reply]
> You can load it over TFTP (or any other supported protocol) as well, http is not a requirement.
> 
> If you have text... Don upload it as an image. Why are you not copy pasting the text?

I don't think you're understanding what you want me to do. I want to add a custom image as a background for the boot menu.

**Reply by NiKiZe** [reply]
Yes I understand that. Can you load the image into iPXE using ingfetch? (You can check result of ingfetch with imgstat)

Because if you can do that then the url you need for the background image is the same. You should even be able to use "logo.png" as that is probably relative to the iPXE script file

So in your script file, add
imgfetch logo.png
imgstat

And pause your script, then post the result (this works fine as a screenshot)

**Reply by davekimi14** [reply]
imgfetch returns okay as well as imgstat looks fine. I then try to load it in using console, to no avail. (did this in the ipxe shell)
![image](https://user-images.githubusercontent.com/46170802/234063329-0a269506-2d55-4a6f-b8ee-bf203cca497e.png)

**Reply by NiKiZe** [reply]
Ah, so you get a error message, that is great debug information...

https://ipxe.org/2d4ce08e

logo.png loads perfectly fine, but switching to graphics mode fails.

**Reply by davekimi14** [reply]
Should I try downscaling my png file down to 1024x768? or is that not part of the issue

**Reply by NiKiZe** [reply]
Did you try setting the display to any graphics mode? (Using the console command)

**Reply by davekimi14** [reply, marked as answer]
Okay I got it figured out. It was the resolution. 
You should set your resolution to what works for you. (mine was 1024 x 768) and then making sure your image is the same resolution.
`console --x 1024 --y 768`
`console --picture logo.png`

**Reply by mcb30** [reply]
If you provide a background picture but do not specify a resolution, then the `console` command will automatically select the smallest mode large enough to display the picture.  In your case, most likely your picture was too big to be displayed in any graphics mode.
